ruby/ql/src/queries/security/cwe-116/IncompleteSanitization.ql

```diff
@@ -0,0 +1,220 @@
+predicate isSimpleAlt(RE::RegExpAlt t) {
+  forall(RE::RegExpTerm ch | ch = t.getAChild() | isSimple(ch))
+}
+
+/**
+ * Holds if `sub` is of the form `x.gsub[!](pattern, replacement)`, where
+ * `pattern` has a known constant value (either a string or a regexp), and
+ * `replacement` prefixes the matched string with a backslash.
+ */
+predicate isBackslashEscape(StringSubstitutionCall sub) {
+  sub.isGlobal() and
+  (exists(sub.getPatternString()) or exists(sub.getPatternRegExp().getRegExpTerm())) and
+  (
+    // Replacement with `\` followed by a backref such as `\&`, `\1`, etc. The
+    // call in Ruby will look something like `str.sub(/.../, '\\\\\0')`. That
+    // replacement string's value (i.e. after Ruby's unescaping) will be `\\\0`.
+    // To account for the backslash escaping in both QL's string syntax and its
+    // regexp engine, each of those three backslashes becomes `\\\\` in the
+    // following:
+    sub.getReplacementString().regexpMatch("\\\\\\\\\\\\(&|\\d)")
+    or
+    // replacement of `c` with `\c`
+    exists(string c | sub.replaces(c, "\\" + c))
+  )
+}
+
+/**
+ * Holds if data flowing into `nd` has no un-escaped backslashes.
+ */
+predicate allBackslashesEscaped(DataFlow::Node node) {
+  exists(StringSubstitutionCall sub | node = sub |
+    // check whether `sub` itself escapes backslashes
+    isBackslashEscape(sub) and
+    (
+      sub.getAReplacedString() = "\\"
+      or
+      // if it's a complex regexp, we conservatively assume that it probably escapes backslashes
+      exists(RE::RegExpTerm root | root = sub.getPatternRegExp().getRegExpTerm() |
+        not isSimple(root)
+      )
+    )
+  )
+  or
+  // flow through string methods
+  exists(ExprNodes::MethodCallCfgNode mc, string m, DataFlow::Node receiver |
+    m =
+      [
+        "sub", "gsub", "slice", "[]", "strip", "lstrip", "rstrip", "chomp", "chop", "downcase",
+        "upcase", "sub!", "gsub!", "slice!", "strip!", "lstrip!", "rstrip!", "chomp!", "chop!",
+        "downcase!", "upcase!",
+      ]
+  |
+    mc = node.asExpr() and
+    m = mc.getExpr().getMethodName() and
+    receiver.asExpr() = mc.getReceiver() and
+    allBackslashesEscaped(receiver)
+  )
+  or
+  // general data flow
+  allBackslashesEscaped(node.getAPredecessor())
+  or
+  // general data flow from a (destructive) [g]sub!
+  exists(DataFlowPrivate::PostUpdateNode post, StringSubstitutionCall sub |
+    sub.isDestructive() and
+    allBackslashesEscaped(sub) and
+    post.getPreUpdateNode() = sub.getReceiver() and
+    post.getASuccessor() = node
+  )
+}
+
+/**
+ * Holds if `sub` looks like a string substitution call that deliberately
+ * removes the first occurrence of `str`.
+ */
+predicate removesFirstOccurence(StringSubstitutionCall sub, string str) {
+  not sub.isGlobal() and sub.replaces(str, "")
+}
+
+/**
+ * Gets a method call where the receiver is the result of a string subtitution
+ * call.
+ */
+DataFlow::Node getAMethodCall(StringSubstitutionCall call) {
```

Comment:
This predicate could perhaps be simplified by having a `DataFlow::CallNode` result type.

Reply:
Indeed it can, thanks.